### DISCUSSION TITLE
"scoped instances" in Microsoft.Extensions.ObjectPool.DependencyInjection

### DISCUSSION THREAD

**Discussion by KalleOlaviNiemitalo** [dotnet/extensions#4515] [Q&A]
The documentation of the Microsoft.Extensions.ObjectPool.ObjectPoolServiceCollectionExtensions.AddPooled methods mentions "scoped instances":

```csharp
    /// <summary>
    /// Adds an <see cref="ObjectPool{TService}"/> and let DI return scoped instances of TService.
    /// </summary>
```

However, this is the only mention of "scoped" in the Microsoft.Extensions.ObjectPool.DependencyInjection project.  Those methods register the ObjectPool\<TService\> service as a singleton; and when the DefaultObjectPool\<TService\> instance is created, its DependencyInjectionPooledObjectPolicy\<TService, TImplementation\> captures a reference to the IServiceProvider, which presumably is the root rather than any IServiceScope.ServiceProvider.  If a scoped service were somehow injected to a pooled object, and the object kept a reference to the service, then it would not be safe to return the pooled object to the singleton pool, where it could outlive the service.

What then is the meaning of "scoped instances" here; is it just a documentation error, or is the implementation incomplete?

**Comment by RussKie** [comment]
@sebastienros @geeknoid can you help?

**Reply by geeknoid** [reply]
@tekian Any feedback?

**Comment by KalleOlaviNiemitalo** [comment]
Re "is the implementation incomplete" — I thought maybe this library defines some class like

```csharp
public sealed class PooledObject<TService> : IDisposable
{
    public TService Instance { get; }
    public void Dispose(); // returns to the pool
}
```

and then registers PooledObject\<TService\> as a scoped service so that it can be injected to constructors in a service scope, and the object is returned to the pool when the scope is disposed.  ObjectPoolServiceCollectionExtensions doesn't register anything like that though, and the library never calls ObjectPool\<TService\>.Get or ObjectPool\<TService\>.Return.

**Reply by KalleOlaviNiemitalo** [reply]
The reference to "scoped instances" was there already in the "Initial commit" on 2023-05-23: <https://github.com/dotnet/extensions/blob/a8c95411ab0ff3895869b2339ce85f573e259d35/src/ToBeMoved/DependencyInjection.Pools/DependencyInjectionExtensions.cs#L23>

That version likewise registered ObjectPool\<T\> and IConfigureOptions\<PoolOptions\> as singletons, did not register any scoped services other than IOptionsSnapshot\<\>, and did not register T itself as a service.

(I'm a bit surprised that this "Initial commit" is newer than the `v5.0.0` and `v2.1.32` tags, and that the current `main` branch has no common history with those tags.)

**Comment by KalleOlaviNiemitalo** [comment]
Maybe it should say

```csharp
    /// <summary>
    /// Adds an <see cref="ObjectPool{TService}"/> from which instances of <typeparamref name="TService"/> can be borrowed.
    /// </summary>
```
